File: projects/5.R
library(keras)
## from https://github.com/rstudio/keras/issues/937
if(FALSE){
  keras::install_keras(version = "2.1.6", tensorflow = "1.5")
}
keras::use_implementation("keras")
keras::use_backend("tensorflow")
future::plan("multiprocess")

## SPAM example.
if(!file.exists("spam.data")){
  download.file(
    "https://web.stanford.edu/~hastie/ElemStatLearn/datasets/spam.data",
    "spam.data")
}

spam.dt <- data.table::fread("spam.data")
label.col <- ncol(spam.dt)
y <- array(spam.dt[[label.col]], nrow(spam.dt))

set.seed(1)
fold.vec <- sample(rep(1:5, l=nrow(spam.dt)))
test.fold <- 1
is.test <- fold.vec == test.fold
is.train <- !is.test

X.sc <- scale(spam.dt[, -label.col, with=FALSE])
X.train.mat <- X.sc[is.train,]
X.test.mat <- X.sc[is.test,]
y.train <- y[is.train]
y.test <- y[is.test]

set.seed(1)
is.subtrain <- sample(rep(c(TRUE, FALSE), l=length(y.train)))
is.validation <- !is.subtrain
X.subtrain.mat <- X.train.mat[is.subtrain,]
X.validation.mat <- X.train.mat[is.validation,]
y.subtrain <- y.train[is.subtrain]
y.validation <- y.train[is.validation]

## analysis of number of hidden units.

units.metrics.list <- list()

(hidden.units.vec <- 2^seq(1, 10))
hidden.units.new <- hidden.units.vec[
  ! hidden.units.vec %in% names(units.metrics.list)]

##for(hidden.units.i in seq_along(hidden.units.new)){
units.metrics.list.new <- future.apply::future_lapply(
  hidden.units.new, function(hidden.units){
    print(hidden.units)
    model <- keras_model_sequential() %>%
      layer_dense(#hidden layer
        input_shape=ncol(X.train.mat),
        units = hidden.units,
        activation = "sigmoid",
        use_bias=FALSE) %>% 
      layer_dense(1, activation = "sigmoid", use_bias=FALSE) #output layer
    model %>%
      compile(
        loss = "binary_crossentropy",
        optimizer = optimizer_adam(lr=0.01),
        metrics = "accuracy" #TODO exact AUC.
      )
    result <- model %>%
      fit(
        x = X.subtrain.mat, y = y.subtrain,
        epochs = 100,
        ##validation_split = 0.3,
        validation_data=list(X.validation.mat, y.validation),
        verbose = 2
      )
    metrics.wide <- do.call(data.table::data.table, result$metrics)
    metrics.wide[, epoch := 1:.N]
    ## units.metrics.list[[paste(hidden.units)]] <- data.table::data.table(
    ##   hidden.units, metrics.wide)
    data.table::data.table(hidden.units, metrics.wide)
  })
units.metrics.list[paste(hidden.units.new)] <- units.metrics.list.new

units.metrics <- do.call(rbind, units.metrics.list)

library(ggplot2)
ggplot()+
  geom_line(aes(
    x=epoch, y=val_loss, color=log10(hidden.units), group=hidden.units),
    data=units.metrics)+
  scale_color_gradient(low="grey", high="red")+
  theme_bw()

ggplot()+
  geom_line(aes(
    x=epoch, y=val_loss),
    data=units.metrics)+
  theme_bw()+
  facet_wrap("hidden.units")

(units.metrics.tall <- nc::capture_melt_single(
  units.metrics,
  set="val_|",
  metric="loss|acc"))
units.metrics.tall[, Set := ifelse(set=="val_", "validation", "subtrain")]

units.loss.tall <- units.metrics.tall[metric=="loss"]
units.loss.min <- units.loss.tall[, .SD[
  which.min(value)], by=.(Set, hidden.units)]
ggplot()+
  geom_line(aes(
    x=epoch, y=value, color=Set),
    data=units.loss.tall)+
  geom_point(aes(
    x=epoch, y=value, color=Set),
    data=units.loss.min)+
  theme_bw()+
  facet_wrap("hidden.units")

ggplot()+
  geom_line(aes(
    x=hidden.units, y=value, color=Set),
    data=units.loss.min)+
  scale_x_log10()


## analyzing number of hidden layers.
layers.metrics.list <- list()

(hidden.layers.vec <- 0:5)
hidden.layers.new <- hidden.layers.vec[
  ! hidden.layers.vec %in% names(layers.metrics.list)]

##for(hidden.layers.i in seq_along(hidden.layers.new)){
layers.metrics.list.new <- future.apply::future_lapply(
  hidden.layers.new, function(hidden.layers){
    print(hidden.layers)
    model <- keras_model_sequential()
    for(layer.i in (0:hidden.layers)[-1]){
      layer_dense(
        model, #hidden layer
        input_shape=ncol(X.train.mat),
        units = 10,
        activation = "relu",
        use_bias=FALSE)
    }
    layer_dense(
      model,
      units=1,
      input_shape=ncol(X.train.mat),
      activation = "sigmoid", use_bias=FALSE) #output layer
    model %>%
      compile(
        loss = "binary_crossentropy",
        optimizer = optimizer_adam(lr=0.001),
        metrics = "accuracy" #TODO exact AUC.
      )
    result <- model %>%
      fit(
        x = X.subtrain.mat, y = y.subtrain,
        epochs = 100,
        ##validation_split = 0.3,
        validation_data=list(X.validation.mat, y.validation),
        verbose = 2
      )
    metrics.wide <- do.call(data.table::data.table, result$metrics)
    metrics.wide[, epoch := 1:.N]
    ## layers.metrics.list[[paste(hidden.layers)]] <- data.table::data.table(
    ##   hidden.layers, metrics.wide)
    data.table::data.table(hidden.layers, metrics.wide)
  })
layers.metrics.list[paste(hidden.layers.new)] <- layers.metrics.list.new

layers.metrics <- do.call(rbind, layers.metrics.list)

(layers.metrics.tall <- nc::capture_melt_single(
  layers.metrics,
  set="val_|",
  metric="loss|acc"))
layers.metrics.tall[, Set := ifelse(set=="val_", "validation", "subtrain")]

layers.loss.tall <- layers.metrics.tall[metric=="loss"]
layers.loss.min <- layers.loss.tall[, .SD[
  which.min(value)], by=.(Set, hidden.layers)]
library(ggplot2)
ggplot()+
  geom_line(aes(
    x=epoch, y=value, color=Set),
    data=layers.loss.tall)+
  geom_point(aes(
    x=epoch, y=value, color=Set),
    data=layers.loss.min)+
  theme_bw()+
  facet_wrap("hidden.layers")

gg <- ggplot()+
  geom_line(aes(
    hidden.layers, value, color=Set),
    data=layers.loss.min)+
  facet_grid(metric ~ .)
directlabels::direct.label(gg, "last.polygons")+
  xlim(0, 6)

File: projects/5.py
import tensorflow as tf
import numpy as np
import plotnine as p9 #conda install -c conda-forge plotnine
import pandas as pd
from metrics import ExactAUC, testN, testConst

spam = np.genfromtxt("spam.data", delimiter=" ")
X_unscaled = spam[:,:-1]
X = (X_unscaled-X_unscaled.mean(axis=0))/X_unscaled.std(axis=0)
y = spam[:,-1]
n_folds = 5
test_fold_ids = np.arange(n_folds)
np.random.seed(0)
test_fold_vec = np.random.permutation(np.tile(test_fold_ids, len(y))[:len(y)])

test_fold = 0
set_dict = {
    "test": test_fold_vec == test_fold,
    "train": test_fold_vec != test_fold,
    }
X_dict = {}
y_dict = {}
for set_name, is_set in set_dict.items():
    X_dict[set_name] = X[is_set, :]
    y_dict[set_name] = y[is_set]

inputs = tf.keras.Input(shape=(X.shape[1],))
hidden = tf.keras.layers.Dense(
    100, activation="sigmoid", use_bias=False)(inputs)
outputs = tf.keras.layers.Dense(
    1, activation="sigmoid", use_bias=False)(hidden)
model = tf.keras.Model(inputs=inputs, outputs=outputs, name="spam_model")

model.compile(
    optimizer=tf.keras.optimizers.Adam(),
    loss=tf.keras.losses.binary_crossentropy,
    metrics=["accuracy", ExactAUC, testN, testConst])

n_epochs = 100
history = model.fit(
    X_dict["train"], y_dict["train"],
    epochs=n_epochs,
    verbose=2,
    validation_split=0.5)
test_scores = dict(zip(
    model.metrics_names,
    model.evaluate(X_dict["test"], y_dict["test"])))
history.history

history_wide = pd.DataFrame(history.history)
history_wide["epoch"] = np.arange(n_epochs)+1
history_tall = pd.melt(history_wide, id_vars="epoch")
history_tall_sets = pd.concat([
    history_tall,
    history_tall["variable"].str.extract("(?P<prefix>val_|)(?P<metric>.*)"),
    ], axis=1)
history_tall_sets["set"] = history_tall_sets["prefix"].apply(
    lambda x: "validation" if x == "val_" else "train")
best_i = history_wide["val_loss"].idxmin()
best_epoch = history_wide.loc[ best_i, "epoch" ]
best_tall = history_tall_sets.loc[ history_tall_sets["epoch"]==best_epoch ]

gg = p9.ggplot(history_tall_sets, p9.aes(x="epoch", y="value", color="set"))+\
    p9.geom_line()+\
    p9.theme_bw()+\
    p9.facet_grid("metric ~ .", scales="free")+\
    p9.geom_point(data=best_tall)+\
    p9.theme(
        facet_spacing={'right': 0.75},
        panel_spacing=0)
gg.save("5-acc-loss.png", width=5, height=5)
